pkg/iam/iam.go

```
@@ -0,0 +1,327 @@
+		return nil, errors.New("cannot create IAMService with empty IAMRoleName")
+	}
+	if config.Log == nil {
+		return nil, errors.New("cannot create IAMService with Log equal to nil")
+	}
+	if !(config.RoleType == ControlPlaneRole || config.RoleType == NodesRole) {
+		return nil, fmt.Errorf("cannot create IAMService with invalid RoleType '%s'", config.RoleType)
+	}
+	client := awsiam.New(config.AWSSession)
+
+	l := config.Log.WithValues("clusterName", config.ClusterName, "iam-role", config.RoleType)
+
+	s := &IAMService{
+		clusterName: config.ClusterName,
+		iamClient:   client,
+		iamRoleName: config.IAMRoleName,
+		log:         l,
+		roleType:    config.RoleType,
+		region:      client.SigningRegion,
+	}
+
+	return s, nil
+}
+
+func (s *IAMService) Reconcile() error {
+	s.log.Info("reconciling IAM role")
+
+	err := s.createMainRole()
+	if err != nil {
+		return err
+	}
+
+	// we only attach the inline policy to a role that is owned (and was created) by iam controller
+	owned, err := isOwnedByIAMController(s.iamRoleName, s.iamClient)
+	if err != nil {
+		s.log.Error(err, "Failed to fetch IAM Role")
+		return err
+	}
+	// check if the policy is created by this controller, if its not than we skip adding inline policy
+	if !owned {
+		s.log.Info("IAM role is not owned by IAM controller, skipping adding inline policy")
+	} else {
+		err = s.attachInlinePolicy()
+		if err != nil {
+			return err
+		}
+	}
+
+	s.log.Info("finished reconciling IAM role")
+	return nil
+}
+
+// createMainRole will create the main IAM role that will be attached to EC2 instances
+func (s *IAMService) createMainRole() error {
+	i := &awsiam.GetRoleInput{
+		RoleName: aws.String(s.iamRoleName),
+	}
+
+	_, err := s.iamClient.GetRole(i)
+
+	// create new IAMRole if it does not exists yet
+	if IsNotFound(err) {
+		assumeRolePolicyDocument, err := generateAssumeRolePolicyDocument(s.region)
+		if err != nil {
+			s.log.Error(err, "failed to generate assume policy document from template for IAM role")
```

Nit: I would say we should either return the error or log the error

---

yeah I am not happy with kubebuilder error logging, as here if I just return error that was generated by the go.template then we will have no idea which template failed and it will make debugging  harder


are we using something like `microerror.Maskf` in kubebuilder or not?

---

I will merge now, but lets chat about this tomorrow in standup